### PR TITLE
Random allocator

### LINKED ISSUE #6: Random allocation algorithm

### PR DESCRIPTION
**Pull request by mdavezac**
Creates a random allocation algorithm. The algorithm accepts any matcher as a constraint, whether these are geographical or technical constraints.

There are currently two limitations, outlined in #21 and #23.

### CODE REVIEW

**Comment by mdavezac** [comment]
@cc-a, @toytun, @DT16, this PR is also ready for review, though it will be easier to read once #12 is merged

**Review by DT16 — COMMENTED** [review]
Looks good. No further comment. As discussed we will work on more sophisticated algorithms later.

File: src/evosim/allocators.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        # TODO: tighten random allocator's default maxiter
+        # labels: enhancement
+        # The default maximum number of iteration is equal to the number of charging
+        # point vacancies. But that could be tightened within the loop itself, whenever
+        # vacancies are allocated. Not sure how though.
+        maxiter = len(vacancies)
+
+    rng.shuffle(vacancies)
+
+    assignement = -np.ones(len(electric_vehicles), dtype=int)
```
Comment (cc-a) [inline src/evosim/allocators.py]:
```suggestion
    assignment = -np.ones(len(electric_vehicles), dtype=int)
```

File: src/evosim/allocators.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    unassigned = electric_vehicles.copy(deep=False)
+    for _ in range(maxiter):
+        is_match = matcher(
+            unassigned.reset_index(drop=True),
+            charging_points.loc[vacancies[: len(unassigned)]].reset_index(drop=True),
+        ).to_numpy()
+        assignement[assignement < 0] = np.where(
```
Comment (cc-a) [inline src/evosim/allocators.py]:
```suggestion
        assignment[assignment < 0] = np.where(
```

File: src/evosim/allocators.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            np.concatenate(
+                (~is_match, np.ones(len(vacancies) - len(is_match), dtype=bool))
+            )
+        ]
+        if len(unassigned) == 0:
+            break
+        vacancies = np.roll(vacancies, shift=1)
+    result = electric_vehicles.copy(deep=False)
+    result["allocation"] = pd.Series(
+        np.where(assignement < 0, pd.NA, assignement), dtype="Int64"
```
Comment (cc-a) [inline src/evosim/allocators.py]:
```suggestion
        np.where(assignment < 0, pd.NA, assignment), dtype="Int64"
```

File: src/evosim/allocators.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        # The default maximum number of iteration is equal to the number of charging
+        # point vacancies. But that could be tightened within the loop itself, whenever
+        # vacancies are allocated. Not sure how though.
+        maxiter = len(vacancies)
+
+    rng.shuffle(vacancies)
+
+    assignement = -np.ones(len(electric_vehicles), dtype=int)
+
+    unassigned = electric_vehicles.copy(deep=False)
```
Comment (cc-a) [inline src/evosim/allocators.py:80]: The below algorithm takes a little while to get your head around. Suggest adding a few comments below for clarity.

File: src/evosim/allocators.rst [hunk 5]
```diff
@@ -0,0 +1,172 @@
+Allocators
+==========
+
+Allocators are algorithms that allocate electric vehicles to charging points. They will
+generally take as input arguments a dataframe of electric vehicles, a dataframe of
```
Comment (cc-a) [inline src/evosim/allocators.rst]:
```suggestion
generally take as input arguments a dataframe of electric vehicles and a dataframe of
```

**Review by cc-a — APPROVED** [review]
Looks good. Assignement should be assignment throughout.

File: src/evosim/allocators.py (hunk 1 above)
Comment (mdavezac) [reply src/evosim/allocators.py]: My French leaking in ;)

File: src/evosim/allocators.py (hunk 4 above)
Comment (mdavezac) [reply src/evosim/allocators.py:80]: added an explanation in the docstring.

**Pull request merged**